### PR TITLE
Initial Greetings with student knowing some English

### PR DESCRIPTION
**Pull request by jsy1218-zz**
I implemented the initial greetings with the scenarios 1) student knows and replies in English 2) student knows English, but replies in native language.

### CODE REVIEW

File: DTML.EduBot/Dialogs/RootDialog.cs [hunk 1]
```diff
@@ -3,76 +3,265 @@
...
+        private Task UserNameReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> result)
+        {
             PromptDialog.Choice(
                 context,
-                this.AfterChoiceSelected,
-                Choices,
-                "Hello Dear " + friendlyUserName + ",\n What would you like to do.",
+                this.AfterDialogChoiceSelectedAsync,
+                DialogChoices,
+                $"Hello Dear {result.GetAwaiter().GetResult().Text},\n What would you like to do.",
```
Comment (sinyaa) [inline DTML.EduBot/Dialogs/RootDialog.cs]: this can throw, needs better error handling

File: DTML.EduBot/Dialogs/RootDialog.cs [hunk 2]
```diff
@@ -3,76 +3,265 @@
...
             return Task.CompletedTask;
         }
 
-        private async Task AfterChoiceSelected(IDialogContext context, IAwaitable<string> result)
+        private async Task UserNameReceivedInNativeLanguageAsync(IDialogContext context, IAwaitable<IMessageActivity> result)
+        {
+            string nativeLanguageIsoCode = MessageTranslator.DEFAULT_LANGUAGE;
+            using (var scope = WebApiApplication.FindContainer().BeginLifetimeScope())
+            {
+                UserData.UserData userData = scope.Resolve<IUserDataRepository>().GetUserData(context.Activity.From.Id);
```
Comment (sinyaa) [inline DTML.EduBot/Dialogs/RootDialog.cs]: Just general comment in IoC using Resolve is a bad practice, we should move either to controller or property injection

File: DTML.EduBot/UserData/UserDataRepository.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Microsoft.Bot.Builder.Internals.Fibers;
+
+namespace DTML.EduBot.UserData
+{
+    public class UserDataRepository : IUserDataRepository
+    {
+        // TODO persist in database instead in memory
```
Comment (sinyaa) [inline DTML.EduBot/UserData/UserDataRepository.cs:10]: We can use Google Firebase for storage its amazing, DTML has an account

File: DTML.EduBot/Dialogs/RootDialog.cs [hunk 4]
```diff
@@ -3,76 +3,265 @@
...
+                {
+                    // detected user language
+
+                    CultureInfo[] allCultures =CultureInfo.GetCultures(CultureTypes.AllCultures);
+                    CultureInfo detectedCulture =
+                       allCultures
+                            .Where(cultureInfo => (detectedLanguageIsoCode.Contains(cultureInfo.TwoLetterISOLanguageName)))
+                            .FirstOrDefault();
+
+                    string detectedLanguageName = CultureInfo.GetCultureInfo(MessageTranslator.DEFAULT_LANGUAGE).DisplayName;
```
Comment (dalin7) [inline DTML.EduBot/Dialogs/RootDialog.cs]: This is getting the language name of DEFAULT_LANGUAGE, not the detected culture. Shouldn't this simply be `detectedCulture.DisplayName`

File: DTML.EduBot/Dialogs/RootDialog.cs [hunk 5]
```diff
@@ -3,76 +3,265 @@
...
+                    
+                    string translatedTooManyAttemptMessage = await MessageTranslator.TranslateTextAsync(TooManyAttemptMessage, userData.NativeLanguageIsoCode);
+
+                    await context.PostAsync($"{translatedTooManyAttemptMessage}");
+                }
+                await Conversation.SendAsync(context.MakeMessage(), () => this);
+            }
+        }
+
+        private async Task<ICollection<string>> TranslatedChoices(IEnumerable<string> choices, string languageToTranslate)
```
Comment (dalin7) [inline DTML.EduBot/Dialogs/RootDialog.cs]: This should probably be part of a different class, maybe the Translator class? We will want all dialogs to be able to have access to this functionality

File: DTML.EduBot/Dialogs/RootDialog.cs [hunk 6]
```diff
@@ -3,76 +3,265 @@
...
+                scope.Resolve<IUserDataRepository>().UpdateUserData(userData);
+
+                if (MessageTranslator.DEFAULT_LANGUAGE.Equals(detectedLanguageIsoCode))
+                {
+                    await context.PostAsync(BotPersonality.UserNameQuestion);
+
+                    // detected it's english language
+                    context.Wait(this.UserNameReceivedAsync);
+                }
+                else
```
Comment (dalin7) [inline DTML.EduBot/Dialogs/RootDialog.cs:67]: Looking throughout this change, I noticed that there are many if/else statements due to language choice as well as different method implementations depending on whether the user is using the default language or not. I realized that we would basically have to do this everywhere else throughout the code.

I think we can avoid this. One solution would be to always call the translate methods and have the translate methods just return the English strings immediately if the language is the default language. Another option that could even avoid those if checks would be to have two classes that implement an interface. We choose which to instantiate based on the language chosen. One class would simply return the strings and the other would do the translation

**Review by dalin7 — CHANGES_REQUESTED** [review]
See comments

File: DTML.EduBot/Dialogs/RootDialog.cs (hunk 6 above)
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/RootDialog.cs:67]: Offline discussed. The if/else statements is due to the scenario user replies in Spanish, but want to keep the conversation in English. Refactoring would require some time and thoughts.

In addition, do we want to refactor during the hackathon, or after?

File: DTML.EduBot/Dialogs/RootDialog.cs (hunk 1 above)
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/RootDialog.cs]: Do you mean result.GetAwaiter().GetResult().Text?

File: DTML.EduBot/Dialogs/RootDialog.cs (hunk 2 above)
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/RootDialog.cs]: Issue created to keep track: https://github.com/eklavyamirani/DTML.EduBot/issues/22

File: DTML.EduBot/Common/MessageTranslator.cs [hunk 7]
```diff
@@ -12,6 +12,7 @@
 
     public class MessageTranslator
     {
+        public static readonly string DEFAULT_LANGUAGE = "en";
```
Comment (eklavyamirani) [inline DTML.EduBot/Common/MessageTranslator.cs:16]: const string

File: DTML.EduBot/DTML.EduBot.csproj [hunk 8]
```diff
@@ -182,6 +186,7 @@
     <None Include="Web.Release.config">
       <DependentUpon>Web.config</DependentUpon>
     </None>
+    <None Include="Web.secrets.Config" />
```
Comment (eklavyamirani) [inline DTML.EduBot/DTML.EduBot.csproj]: Don't include.

File: DTML.EduBot/Dialogs/RootDialog.cs [hunk 9]
```diff
@@ -3,76 +3,265 @@
...
+        private const string No = "No";
+
+        private static readonly IEnumerable<string> YesNoChoices = new ReadOnlyCollection<string>
+            (new List<String> {
+                Yes,
+                No});
+
         private const string ChatWithBot = "Chat With Bot";
-        private const string StartTheLessonPlan = "Start The Lesson Plan";
+        private const string StartTheLessonPlan = "Start English Lesson Plan";
```
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/RootDialog.cs]: Move all strings to Constants file.

File: DTML.EduBot/Dialogs/RootDialog.cs [hunk 10]
```diff
@@ -3,76 +3,265 @@
...
         {
             context.Wait(this.MessageReceivedAsync);
             return Task.CompletedTask;
         }
 
-        private Task MessageReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> result)
+        private async Task MessageReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> result)
         {
-            string friendlyUserName = context.Activity.From.Name;
+            var userText = (await result).Text;
```
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/RootDialog.cs]: possible null reference. These will creep in a demo unexpectedly.

File: DTML.EduBot/Dialogs/RootDialog.cs (hunk 1 above)
Comment (eklavyamirani) [reply DTML.EduBot/Dialogs/RootDialog.cs]: yes. for example task canceled exception.

File: DTML.EduBot/Dialogs/RootDialog.cs [hunk 11]
```diff
@@ -3,76 +3,265 @@
...
+                attempts: MAX_ATTEMPT);
+        }
+
+        private async Task AfterDialogChoiceSelectedAsync(IDialogContext context, IAwaitable<string> result)
         {
             try
             {
                 var selection = await result;
 
+                // TODO: enum.
```
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/RootDialog.cs]: lol. Too many todos for enums.

**Review by eklavyamirani — APPROVED** [review]
Address review comments before merging.

File: DTML.EduBot/Dialogs/RootDialog.cs (hunk 6 above)
Comment (vyjayanthiraja) [reply DTML.EduBot/Dialogs/RootDialog.cs:67]: +1 we should create an issue to track this change!

**Pull request merged**